### PR TITLE
Extend CheckpointSaver class

### PR DESCRIPTION
**Pull request by nicofirst1**
Extended the CheckpointSaver to add the following:
- possibility to define the starting epoch. This comes in handy when you do not want to use the Trainer.load_checkpoint.
- Define a max checkpoint files to keep at the same time. For big enough model and many epochs the checkpoint directory may become saturated with too many heavy files. With this functionality a max number of files are allowed in the directory at the same time. Oldest checkpoints are deleted first.

### CODE REVIEW

**Comment by nicofirst1** [comment]
I cannot replicate the error on my machine.
@robertodessi can you take a look?

File: egg/core/callbacks.py [hunk 1]
```diff
@@ -131,11 +134,23 @@ def __init__(
         checkpoint_path: Union[str, pathlib.Path],
         checkpoint_freq: int = 1,
         prefix: str = "",
+        start_epoch: int = 0,
+        max_chk: int = sys.maxsize,
     ):
+        """Saves a checkpoint file for training
+        Arguments:
+            checkpoint_path {Union[str, pathlib.Path]} -- path to checkpoint directory, will be created if not present
```
Comment (robertodessi) [inline egg/core/callbacks.py:142]: Please use [this](https://github.com/facebookresearch/EGG/blob/master/egg/core/reinforce_wrappers.py#L621-L629) style (with a full stop at the end of method description).

File: egg/core/callbacks.py [hunk 2]
```diff
@@ -131,11 +134,23 @@ def __init__(
...
+            prefix {str} -- Name of checkpoint file, will be {prefix}{current_epoch}.tar
+            start_epoch {int} -- For resuming training you can either use Trainer.load_from_latest or specify the epoch
+             here
+            max_chk {int} -- Max number of concurrent checkpoint files in the directory.
+        """
         self.checkpoint_path = pathlib.Path(checkpoint_path)
         self.checkpoint_freq = checkpoint_freq
         self.prefix = prefix
-        self.epoch_counter = 0
+        self.epoch_counter = start_epoch
```
Comment (robertodessi) [inline egg/core/callbacks.py:152]: This is overwritten at each `on_epoch_end` call and I don't think we need it as a class field. I would use (as somehow you did) the epoch parameter passed to the method at each call given that `Trainer` handle the right epoch number both when training from scratch and when resuming with a trained model (done in `load_from_latest`

File: egg/core/callbacks.py [hunk 3]
```diff
@@ -169,6 +186,30 @@ def get_checkpoint(self):
             optimizer_state_dict=self.trainer.optimizer.state_dict(),
         )
 
+    def get_chk_files(self):
```
Comment (robertodessi) [inline egg/core/callbacks.py:189]: What do you think about using _checkpoint_ instead of _chk_? It somehow confuses me with _chunk_

**Comment by robertodessi** [comment]
Test don't fail now. Except for the comments it looks good to me, thanks!

File: egg/core/callbacks.py (hunk 3 above)
Comment (nicofirst1) [reply egg/core/callbacks.py]: done in 6e029ca

File: egg/core/callbacks.py (hunk 2 above)
Comment (nicofirst1) [reply egg/core/callbacks.py]: removed in 6e029ca

File: egg/core/callbacks.py (hunk 1 above)
Comment (nicofirst1) [reply egg/core/callbacks.py]: fixed in 6e029ca

File: egg/core/callbacks.py [hunk 4]
```diff
@@ -131,10 +134,18 @@ def __init__(
...
         prefix: str = "",
+        max_checkpoints: int = sys.maxsize,
     ):
+        """Saves a checkpoint file for training.
+            :param checkpoint_path:  path to checkpoint directory, will be created if not present
+            :param checkpoint_freq:  Number of epochs for checkpoint saving
+            :param prefix: Name of checkpoint file, will be {prefix}{current_epoch}.tar
+            :param max_checkpoints: Max number of concurrent checkpoint files in the directory.
+        """
         self.checkpoint_path = pathlib.Path(checkpoint_path)
```
Comment (robertodessi) [inline egg/core/callbacks.py:145]:
```suggestions
    self.checkpoint_path = pathlib.Path(checkpoint_path).resolve()
```

File: egg/core/callbacks.py [hunk 5]
```diff
@@ -169,6 +182,30 @@ def get_checkpoint(self):
...
+    def natural_sort(to_sort):
+        """
+        Sort a list of files naturally
+        E.g. [file1,file4,file32,file2] -> [file1,file2,file4,file32]
+        """
+        convert = lambda text: int(text) if text.isdigit() else text.lower()
+        alphanum_key = lambda key: [convert(c) for c in re.split("([0-9]+)", key)]
+        return sorted(to_sort, key=alphanum_key)
+
+    def remove_old_chk(self):
```
Comment (robertodessi) [inline egg/core/callbacks.py:201]:
```suggestion
    def remove_old_checkpoint(self):
```

File: egg/core/callbacks.py [hunk 6]
```diff
@@ -159,6 +170,8 @@ def save_checkpoint(self, filename: str):
         Saves the game, agents, and optimizer states to the checkpointing path under `<number_of_epochs>.tar` name
         """
         self.checkpoint_path.mkdir(exist_ok=True, parents=True)
+        if len(self.get_checkpoint_files()) > self.max_checkpoints_keep:
+            self.remove_old_chk()
```
Comment (robertodessi) [inline egg/core/callbacks.py:174]: remove_oldest_checkpoint

File: egg/core/callbacks.py [hunk 7]
```diff
@@ -169,6 +182,30 @@ def get_checkpoint(self):
...
+
+    @staticmethod
+    def natural_sort(to_sort):
+        """
+        Sort a list of files naturally
+        E.g. [file1,file4,file32,file2] -> [file1,file2,file4,file32]
+        """
+        convert = lambda text: int(text) if text.isdigit() else text.lower()
+        alphanum_key = lambda key: [convert(c) for c in re.split("([0-9]+)", key)]
+        return sorted(to_sort, key=alphanum_key)
```
Comment (robertodessi) [inline egg/core/callbacks.py:199]: if you resume a training there might be a checkpoint ending with _final_, see [here](https://github.com/facebookresearch/EGG/blob/6e029ca514d5f0df102ff45012698caf0b6dedf9/egg/core/callbacks.py#L163-L166)

File: egg/core/callbacks.py [hunk 8]
```diff
@@ -131,10 +134,18 @@ def __init__(
...
+        """Saves a checkpoint file for training.
+            :param checkpoint_path:  path to checkpoint directory, will be created if not present
+            :param checkpoint_freq:  Number of epochs for checkpoint saving
+            :param prefix: Name of checkpoint file, will be {prefix}{current_epoch}.tar
+            :param max_checkpoints: Max number of concurrent checkpoint files in the directory.
+        """
         self.checkpoint_path = pathlib.Path(checkpoint_path)
         self.checkpoint_freq = checkpoint_freq
         self.prefix = prefix
+        self.max_checkpoints_keep = max_checkpoints
```
Comment (robertodessi) [inline egg/core/callbacks.py:148]:
```suggestion
    self.max_checkpoints = max_checkpoints
````

File: egg/core/callbacks.py [hunk 9]
```diff
@@ -159,6 +170,8 @@ def save_checkpoint(self, filename: str):
         Saves the game, agents, and optimizer states to the checkpointing path under `<number_of_epochs>.tar` name
         """
         self.checkpoint_path.mkdir(exist_ok=True, parents=True)
+        if len(self.get_checkpoint_files()) > self.max_checkpoints_keep:
```
Comment (robertodessi) [inline egg/core/callbacks.py:173]:
```suggestions
        if len(self.get_checkpoint_files()) > self.max_checkpoints:
```

File: egg/core/callbacks.py [hunk 10]
```diff
@@ -169,6 +182,30 @@ def get_checkpoint(self):
...
+        alphanum_key = lambda key: [convert(c) for c in re.split("([0-9]+)", key)]
+        return sorted(to_sort, key=alphanum_key)
+
+    def remove_old_chk(self):
+        """
+        Remove the oldest checkpoint from the dir
+        """
+        chk_points = self.natural_sort(self.get_checkpoint_files())
+        to_remove = chk_points[0]
+        os.remove(os.path.join(self.checkpoint_path, to_remove))
```
Comment (robertodessi) [inline egg/core/callbacks.py:207]:
```suggestion
        checkpoints = self.natural_sort(self.get_checkpoint_files())
        os.remove(os.path.join(self.checkpoint_path, checkpoints[0])
````

File: tests/test_trainer.py [hunk 11]
```diff
@@ -97,16 +97,16 @@ def test_snapshoting():
...
-    assert (CHECKPOINT_PATH / Path("2.tar")).exists()
+    trainer.train(n_epochs=6)
+    assert (CHECKPOINT_PATH / Path("5.tar")).exists()
+    assert (CHECKPOINT_PATH / Path("6.tar")).exists()
     assert (CHECKPOINT_PATH / Path("final.tar")).exists()
     del trainer
     trainer = core.Trainer(game, optimizer, train_data=data)  # Re-instantiate trainer
     trainer.load_from_latest(CHECKPOINT_PATH)
-    assert trainer.start_epoch == 2
+    assert trainer.start_epoch == 6
```
Comment (robertodessi) [inline tests/test_trainer.py:109]: how about an additional check that *only* checkpoint_5 and 6 are present in the folder but in a new test?

**Comment by robertodessi** [comment]
Could you please update the method the that is uses epoch directly instead of self.epoch_counter? https://github.com/facebookresearch/EGG/pull/187/files#diff-b377660d119e5f458c4c12c9603a825095d82df80735f903b8902e35dad5217cL141-L149
Thanks!!

File: egg/core/callbacks.py (hunk 6 above)
Comment (robertodessi) [reply egg/core/callbacks.py:174]: and self.max_checkpoints instead of self.max_checkpoints_keep

File: egg/core/callbacks.py (hunk 7 above)
Comment (nicofirst1) [reply egg/core/callbacks.py:199]: the final version is the first of the sorted list

File: tests/test_trainer.py (hunk 11 above)
Comment (nicofirst1) [reply tests/test_trainer.py:109]: How do you do that?

Comment (robertodessi) [reply tests/test_trainer.py:109]: You could leave the task as it was (without max_checkpoints and with two epochs/checkpoints only) and have another test that specifically check the max_checkpoints methods (you can use the current new test with max_checkpoints)

File: egg/core/callbacks.py [hunk 12]
```diff
@@ -131,22 +134,24 @@ def __init__(
...
+        :param checkpoint_path:  path to checkpoint directory, will be created if not present
+        :param checkpoint_freq:  Number of epochs for checkpoint saving
+        :param prefix: Name of checkpoint file, will be {prefix}{current_epoch}.tar
+        :param max_checkpoints: Max number of concurrent checkpoint files in the directory.
+        """
         self.checkpoint_path = pathlib.Path(checkpoint_path)
         self.checkpoint_freq = checkpoint_freq
         self.prefix = prefix
+        self.max_checkpoints = max_checkpoints
         self.epoch_counter = 0
```
Comment (robertodessi) [inline egg/core/callbacks.py:149]: Could you please remove this? I don't think we really need. The trainer instance has the epoch counter anyway

File: egg/core/callbacks.py [hunk 13]
```diff
@@ -131,22 +134,24 @@ def __init__(
...
 
     def on_epoch_end(self, loss: float, logs: Interaction, epoch: int):
         self.epoch_counter = epoch
```
Comment (robertodessi) [inline egg/core/callbacks.py:152]: same here

File: egg/core/callbacks.py [hunk 14]
```diff
@@ -159,6 +164,8 @@ def save_checkpoint(self, filename: str):
         Saves the game, agents, and optimizer states to the checkpointing path under `<number_of_epochs>.tar` name
         """
         self.checkpoint_path.mkdir(exist_ok=True, parents=True)
+        if len(self.get_checkpoint_files()) > self.max_checkpoints:
```
Comment (robertodessi) [inline egg/core/callbacks.py:167]: This should be done only if `max_checkpoints > 0`. If you think at the very corner case where you are resuming a training, by mistake (maybe forgot to update your code) you set the checkpoint path and have `max_checkpoints = 0` then you'd be removing old checkpoints from the previous run

File: egg/core/callbacks.py [hunk 15]
```diff
@@ -169,6 +176,29 @@ def get_checkpoint(self):
...
+        convert = lambda text: int(text) if text.isdigit() else text.lower()
+        alphanum_key = lambda key: [convert(c) for c in re.split("([0-9]+)", key)]
+        return sorted(to_sort, key=alphanum_key)
+
+    def remove_oldest_checkpoint(self):
+        """
+        Remove the oldest checkpoint from the dir
+        """
+        checkpoints = self.natural_sort(self.get_checkpoint_files())
+        os.remove(os.path.join(self.checkpoint_path, checkpoints[0]))
```
Comment (robertodessi) [inline egg/core/callbacks.py:200]:
```suggestion
        if checkpoints:        
            os.remove(os.path.join(self.checkpoint_path, checkpoints[0]))
```

File: egg/core/callbacks.py (hunk 12 above)
Comment (nicofirst1) [reply egg/core/callbacks.py:149]: Do you mean the epoch_counter?
it is used when getting the checkpoint [here](https://github.com/facebookresearch/EGG/blob/edde9f5628dde520bd58f728991091bb65d51f6b/egg/core/callbacks.py#L174)

Comment (nicofirst1) [reply egg/core/callbacks.py:149]: On the other hand if you are referring to max_checkpoints it is needed to keep track of the max number of checkpoints to save

File: egg/core/callbacks.py (hunk 14 above)
Comment (nicofirst1) [reply egg/core/callbacks.py:167]: Well yes, if the number of chkps is grater than max_chkps then the old one must be removed

File: egg/core/callbacks.py (hunk 12 above)
Comment (robertodessi) [reply egg/core/callbacks.py:149]: Sorry my bad, you're right!

File: egg/core/callbacks.py (hunk 14 above)
Comment (robertodessi) [reply egg/core/callbacks.py:167]: This is very very minor though, you can ignore it for the sake of the PR, up to you

File: tests/test_trainer.py (hunk 11 above)
Comment (nicofirst1) [reply tests/test_trainer.py]: done in 
dc6fd3a

Comment (robertodessi) [reply tests/test_trainer.py]: The latest test is just a copy-paste of the the 2-epoch one but checking that even when using max_checkpoints training with 6-epochs the model stores checkpoint_{5, 6, final}. An ideal test would be checking that *only* those checkpoints are present. Thus making sure the the others are correctly deleted

File: egg/core/callbacks.py (hunk 14 above)
Comment (robertodessi) [reply egg/core/callbacks.py:167]: It depends if you consider ma_checkpoints per run (then you shouldn't change them) or max_checkpoints per folder? I think we should max per run. If I set max_checkpoints to 0 it should mean I don't want any checkpoint to be saved in this run, the expected behavior is not deleting checkpoints from a previous run

Let's leave it as is for now though

**Pull request merged**